gemini/use-cases/retrieval-augmented-generation/developer_code_chat/utils/generate_embeddings_utils.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2023 Google LLC. This software is provided as-is, without warranty
+# or representation for any use or purpose. Your use of it is subject to your
+# agreement with Google.
+"""Embedding Utils"""
+
+
+import time
+from typing import List
+
+from langchain.embeddings import VertexAIEmbeddings
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+
+
+def split_documents(documents, split_document_flag="PAGES"):
+    """Split documents either by Pages or by the chunks"""
+
+    # by default documents are split by pages using PyPDFLoader
+    # in GCSDirectoryLoader_modified class
+    doc_splits = documents
+
+    if split_document_flag == "CHUNKS":
+        # split the documents into chunks
+        text_splitter = RecursiveCharacterTextSplitter(
+            chunk_size=1000,
+            chunk_overlap=50,
+            separators=["\n\n", "\n", ".", "!", "?", ",", " ", ""],
+        )
+        doc_splits = text_splitter.split_documents(documents)
+
+    # Add chunk number to metadata
+    for idx, split in enumerate(doc_splits):
+        split.metadata["chunk"] = idx
+
+    return doc_splits
+
+
+def rate_limit(max_per_minute):
+    """Utility functions for Embeddings API with rate limiting"""
+
+    period = 60 / max_per_minute
+    # print("Waiting")
+    while True:
+        before = time.time()
+        yield
+        after = time.time()
+        elapsed = after - before
+        sleep_time = max(0, period - elapsed)
+        if sleep_time > 0:
+            print(".", end="")
+            time.sleep(sleep_time)
+
+
+class CustomVertexAIEmbeddings(VertexAIEmbeddings):
```

Comment:
Can this use the existing VertexAIEmbeddings connector in LangChain? It could be good to make a Pull Request to add the rate limiting if needed.